lib/Dialect/Comb/CombFolds.cpp

```diff
@@ -1116,6 +1117,253 @@ static bool canonicalizeOrOfConcatsWithCstOperands(OrOp op, size_t concatIdx1,
+
+        APInt lhsBound, rhsBound;
+        if (matchPattern(lhsOp.rhs(), m_RConstant(lhsBound)) &&
+            matchPattern(rhsOp.rhs(), m_RConstant(rhsBound))) {
+          ICmpPredicate lhsPred = lhsOp.predicate();
+          ICmpPredicate rhsPred = rhsOp.predicate();
+          if (lhsPred == ICmpPredicate::ugt && rhsPred == ICmpPredicate::ult) {
+            argChecks[arg].emplace_back(i,
+                RangeCheck{lhsBound, rhsBound, lhsOp, rhsOp});
+            continue;
+          }
+          if (lhsPred == ICmpPredicate::ult && rhsPred == ICmpPredicate::ugt) {
+            argChecks[arg].emplace_back(i,
+                RangeCheck{rhsBound, lhsBound, rhsOp, lhsOp});
+            continue;
+          }
+        }
+      }
+    }
+
+    if (auto cmpOp = input.getDefiningOp<ICmpOp>()) {
+      APInt value;
+      if (matchPattern(cmpOp.rhs(), m_RConstant(value))) {
+        // Find equality tests: x == n
+        if (cmpOp.predicate() == ICmpPredicate::eq) {
+          argChecks[cmpOp.lhs()].emplace_back(i, EqCheck{value, cmpOp});
+          continue;
+        }
+        // Find upper bound tests: x < n
+        if (cmpOp.predicate() == ICmpPredicate::ult) {
+          argChecks[cmpOp.lhs()].emplace_back(i, BoundCheck{value, cmpOp});
+          continue;
+        }
+      }
+    }
+
+    keptOperands[i] = true;
+    ++numKeptOperands;
+  }
+
+  // Find an 'n' such that 'n < x' for any 'x' matching the check. If 0 is part
+  // of any of the checks, nullopt is returned to represent -1.
+  auto getLowerBound = [](const AnyCheck &check) -> std::optional<APInt> {
+    if (auto eq = std::get_if<EqCheck>(&check)) {
+      if (eq->value.isZero())
+        return std::nullopt;
+      return eq->value - 1;
+    }
+    if (auto range = std::get_if<RangeCheck>(&check))
+      return range->start;
+    if (auto bound = std::get_if<BoundCheck>(&check))
+      return std::nullopt;
+    llvm_unreachable("unknown check");
+  };
+
+  // Find an 'n' such that 'x < n' for any 'x' matching the check.
+  auto getUpperBound = [](const AnyCheck &check) -> APInt {
+    if (auto eq = std::get_if<EqCheck>(&check))
+      return eq->value + 1;
+    if (auto range = std::get_if<RangeCheck>(&check))
+      return range->end;
+    if (auto bound = std::get_if<BoundCheck>(&check))
+      return bound->value;
+    llvm_unreachable("unknown check");
+  };
+
+  // For each value, try to compress the associated checks.
+  llvm::SmallVector<Value> newOperands;
+  for (auto &[arg, checks] : argChecks) {
+    // Order the checks by their lower bounds.
+    std::stable_sort(checks.begin(), checks.end(), [&](auto &lhs, auto &rhs) {
+      if (auto lowerLHS = getLowerBound(lhs.second))
+        if (auto lowerRHS = getLowerBound(rhs.second))
+          return lowerLHS->ult(*lowerRHS);
+      return false;
+    });
+
+    // Find sequences of overlapping checks and compress them.
+    for (auto it = checks.begin(); it != checks.end();) {
+      auto begin = it++;
```

Comment:
nit: the LLVM preferred style is to always use preincrement: https://llvm.org/docs/CodingStandards.html#prefer-preincrement

Reply:
Here I actually do want to create a copy and increment afterwards. It seems like this pattern is used in other parts of CIRCT as well.

Reply:
llvm::preincrement_iterator?